# Expanding Our Dataset with the Tolnay-Beck Inventory

In [1]:
import pandas as pd
import re
import numpy as np

In [2]:
df = pd.read_excel('tolnay_beck_data_2022.xls')

In [3]:
df

Status  Year  Month Day                    Name Alt Name #1  \
0     Probable lynching  1865      2   0  Unnamed Person #1 of 4         NaN   
1     Probable lynching  1865      2   0  Unnamed Person #2 of 4         NaN   
2     Probable lynching  1865      2   0  Unnamed Person #3 of 4         NaN   
3     Probable lynching  1865      2   0  Unnamed Person #4 of 4         NaN   
4              Lynching  1865      3   5                  — Hall         NaN   
...                 ...   ...    ...  ..                     ...         ...   
5866           Lynching  1981      3  21          Michael Donald         NaN   
5867           Lynching  1998      6   7          James Byrd Jr.         NaN   
5868  Possible lynching  2003      8  17            Kareem Cowan         NaN   
5869  Possible lynching  2011      6  26    James Craig Anderson         NaN   
5870           Lynching  2020      2  23   Ahmaud Marquez Arbery         NaN   

     Alt Name #2 Alt Name #3   Order Victim's Race  ...  \
0            NaN         NaN  1 of 4   White-Anglo  ...   
1            NaN         NaN  2 of 4   White-Anglo  ...   
2            NaN         NaN  3 of 4   White-Anglo  ...   
3            NaN         NaN  4 of 4   White-Anglo  ...   
4            NaN         NaN  1 of 2   White-Anglo  ...   
...          ...         ...     ...           ...  ...   
5866         NaN         NaN  1 of 1         Black  ...   
5867         NaN         NaN  1 of 1         Black  ...   
5868         NaN         NaN  1 of 1         Black  ...   
5869         NaN         NaN  1 of 1         Black  ...   
5870         NaN         NaN  1 of 1         Black  ...   

                PostMort Desc Problem Flag  \
0               None reported          NaN   
1               None reported          NaN   
2               None reported          NaN   
3               None reported          NaN   
4               None reported          NaN   
...                       ...          ...   
5866  Body hanged for display          NaN   
5867            None reported          NaN   
5868            None reported          NaN   
5869            None reported          NaN   
5870            None reported          NaN   

                                        Resident Status Time of Lynching  \
0     Quite possibly strangers. Characterized as “de...              NaN   
1     Quite possibly strangers. Characterized as “de...              NaN   
2     Quite possibly strangers. Characterized as “de...              NaN   
3     Quite possibly strangers. Characterized as “de...              NaN   
4            Strangers. Deserters from Confederate army            Night   
...                                                 ...              ...   
5866  No indication of being a stranger or marginalized              NaN   
5867  No indication of being a stranger or marginalized            Night   
5868  No indication of being a stranger or marginali...              NaN   
5869  No indication of being a stranger or marginalized              NaN   
5870  No indication of being a stranger or marginalized              NaN   

                                             TolnayBeck  \
0      Tolnay-Beck inventory covers the years 1882-1930   
1      Tolnay-Beck inventory covers the years 1882-1930   
2      Tolnay-Beck inventory covers the years 1882-1930   
3      Tolnay-Beck inventory covers the years 1882-1930   
4      Tolnay-Beck inventory covers the years 1882-1930   
...                                                 ...   
5866  Original Tolnay-Beck inventory covers the year...   
5867  Original Tolnay-Beck inventory did not include...   
5868  Original Tolnay-Beck inventory covers the year...   
5869  Original Tolnay-Beck inventory covers the year...   
5870  Original Tolnay-Beck inventory covers the year...   

                     Torture Torture Codes  \
0              None reported             0   
1              None reported             0   
2              None reported      

In [4]:
black_victims = df[df["Victim's Race"].str.contains(r'Black', regex=True) & ~df['Name'].str.contains(r'Unnamed', regex=True, case=False, na=False)]

In [5]:
black_victims['Name'] = black_victims['Name'].str.lower()
black_victims

/var/folders/4l/trq3vqp90ybdgf6j37wrxx440000gn/T/ipykernel_88815/3359694536.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  black_victims['Name'] = black_victims['Name'].str.lower()


Status  Year  Month Day                   Name Alt Name #1  \
6              Lynching  1865      7  20        charles winters         NaN   
7              Lynching  1865      7  20          thos. bradley         NaN   
8              Lynching  1865      7  21             john hirst         NaN   
9              Lynching  1865      7  21         john middleton         NaN   
17             Lynching  1865     11  22                  kit —         NaN   
...                 ...   ...    ...  ..                    ...         ...   
5866           Lynching  1981      3  21         michael donald         NaN   
5867           Lynching  1998      6   7         james byrd jr.         NaN   
5868  Possible lynching  2003      8  17           kareem cowan         NaN   
5869  Possible lynching  2011      6  26   james craig anderson         NaN   
5870           Lynching  2020      2  23  ahmaud marquez arbery         NaN   

     Alt Name #2 Alt Name #3   Order Victim's Race  ...  \
6            NaN         NaN  2 of 2         Black  ...   
7            NaN         NaN  1 of 2         Black  ...   
8            NaN         NaN  1 of 2         Black  ...   
9            NaN         NaN  2 of 2         Black  ...   
17           NaN         NaN  1 of 1         Black  ...   
...          ...         ...     ...           ...  ...   
5866         NaN         NaN  1 of 1         Black  ...   
5867         NaN         NaN  1 of 1         Black  ...   
5868         NaN         NaN  1 of 1         Black  ...   
5869         NaN         NaN  1 of 1         Black  ...   
5870         NaN         NaN  1 of 1         Black  ...   

                PostMort Desc Problem Flag  \
6               None reported          NaN   
7               None reported          NaN   
8               None reported          NaN   
9               None reported          NaN   
17              None reported          NaN   
...                       ...          ...   
5866  Body hanged for display          NaN   
5867            None reported          NaN   
5868            None reported          NaN   
5869            None reported          NaN   
5870            None reported          NaN   

                                        Resident Status Time of Lynching  \
6     No indication of being a stranger or marginalized            Night   
7     No indication of being a stranger or marginalized            Night   
8     No indication of being a stranger or marginalized            Night   
9     No indication of being a stranger or marginalized            Night   
17    No indication of being a stranger or marginalized              NaN   
...                                                 ...              ...   
5866  No indication of being a stranger or marginalized              NaN   
5867  No indication of being a stranger or marginalized            Night   
5868  No indication of being a stranger or marginali...              NaN   
5869  No indication of being a stranger or marginalized              NaN   
5870  No indication of being a stranger or marginalized              NaN   

                                             TolnayBeck  \
6     Original Tolnay-Beck inventory covers the year...   
7     Original Tolnay-Beck inventory covers the year...   
8     Original Tolnay-Beck inventory covers the year...   
9     Original Tolnay-Beck inventory covers the year...   
17    Original Tolnay-Beck inventory covers the year...   
...                                                 ...   
5866  Original Tolnay-Beck inventory covers the year...   
5867  Original Tolnay-Beck inventory did not include...   
5868  Original Tolnay-Beck inventory covers the year...   
5869  Original Tolnay-Beck inventory covers the year...   
5870  Original Tolnay-Beck inventory covers the year...   

                     Torture Torture Codes  \
6              None reported             0   
7              None reported             0   
8              None reported             0   


OOH boy, this will add as much as 3981 victims to our data... The Seguin/Rigby data resulted in less than 350 victims. No doubt this will improve our data, but it's also going to take a long time to scrape all those results from Chron Am. For reference, it took about 9 hours to scrape the 350 results from Seguin & Rigby. Is it going to take like 100 hours to scrape these search results??? Jesus...

Also, an important note for down the line: cities are not mentioned in the Tolnay-Beck inventory. I'll have to geocode these results using the county and state.

Should I go back and geocode Seguin/Rigby by county/state? This would add some congruity to our data. It may also make finding lynching town newspaper data a little easier since I'm pretty sure Chron Am tracks papers by county, too.

Anyway, let's keep refining this data.

In [6]:
black_victims = black_victims[black_victims['Year'] <= 1921]

def remove_punctuation(text):
    return re.sub(r'[^\w\s]', ' ', text)

black_victims['Name'] = black_victims['Name'].apply(remove_punctuation)

def remove_double_space(text):
    return re.sub(r'  ', ' ', text)

black_victims['Name'] = black_victims['Name'].apply(remove_double_space)

black_victims = black_victims[black_victims['Name'].str.split().str.len() > 1]

black_victims

/var/folders/4l/trq3vqp90ybdgf6j37wrxx440000gn/T/ipykernel_88815/3722398447.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  black_victims['Name'] = black_victims['Name'].apply(remove_punctuation)
/var/folders/4l/trq3vqp90ybdgf6j37wrxx440000gn/T/ipykernel_88815/3722398447.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  black_victims['Name'] = black_victims['Name'].apply(remove_double_space)


Status  Year  Month Day              Name        Alt Name #1  \
6              Lynching  1865      7  20   charles winters                NaN   
7              Lynching  1865      7  20      thos bradley                NaN   
8              Lynching  1865      7  21        john hirst                NaN   
9              Lynching  1865      7  21    john middleton                NaN   
22             Lynching  1866      2  16      joseph balls       Joseph Rails   
...                 ...   ...    ...  ..               ...                ...   
5445           Lynching  1921     12   4         west hale           Wes Hale   
5446           Lynching  1921     12  11        fred rouse           Ed Rouse   
5447  Probable lynching  1921     12  11    lonnie newsome     Alonzo Newsome   
5449           Lynching  1921     12  16  lee robinson jr   Joe Robertson Jr.   
5452           Lynching  1921     12  28   bill mcallister               Ed —   

           Alt Name #2 Alt Name #3   Order Victim's Race  ...  PostMort Desc  \
6                  NaN         NaN  2 of 2         Black  ...  None reported   
7                  NaN         NaN  1 of 2         Black  ...  None reported   
8                  NaN         NaN  1 of 2         Black  ...  None reported   
9                  NaN         NaN  2 of 2         Black  ...  None reported   
22        Joseph Ralls         NaN  1 of 2         Black  ...  None reported   
...                ...         ...     ...           ...  ...            ...   
5445               NaN         NaN  3 of 3         Black  ...  None reported   
5446               NaN         NaN  1 of 1         Black  ...  None reported   
5447               NaN         NaN  1 of 1         Black  ...  None reported   
5449  Leo Robinson Jr.         NaN  1 of 1         Black  ...  None reported   
5452               NaN         NaN  1 of 1         Black  ...  None reported   

     Problem Flag                                    Resident Status  \
6             NaN  No indication of being a stranger or marginalized   
7             NaN  No indication of being a stranger or marginalized   
8             NaN  No indication of being a stranger or marginalized   
9             NaN  No indication of being a stranger or marginalized   
22            NaN  No indication of being a stranger or marginali...   
...           ...                                                ...   
5445          NaN  No indication of being a stranger or marginalized   
5446          NaN  No indication of being a stranger or marginali...   
5447          NaN  No indication of being a stranger or marginalized   
5449          NaN  No indication of being a stranger or marginali...   
5452          NaN  No indication of being a stranger or marginali...   

     Time of Lynching                                         TolnayBeck  \
6               Night  Original Tolnay-Beck inventory covers the year...   
7               Night  Original Tolnay-Beck inventory covers the year...   
8               Night  Original Tolnay-Beck inventory covers the year...   
9               Night  Original Tolnay-Beck inventory covers the year...   
22              Night  Original Tolnay-Beck inventory covers the year...   
...               ...                                                ...   
5445              NaN  West Hale, 12/4/1921, Oconee Co., Black Male, ...   
5446         11:00 PM  Original Tolnay-Beck inventory did not include...   
5447            Night  Original Tolnay-Beck inventory did not include...   
5449              NaN  Lee Robinson, Jr., 12/21/1921, Brooks Co., Bla...   
5452            Night  Unnamed Negro, 1/4/1922, Florence Co., Black M...   

            Torture Torture Codes  \
6     None reported             0   
7     None reported             0   
8     None reported             0   
9     None reported             0   
22    None reported             0   
...             ...           ...   
5445  None reported             0   
5446     

Above I've subset the data by the year range relevant to VRT (1865 to 1921). I've I removed punctuation from victim names and removed instances where the victim name was only one name since scraping by individual names results in many false positives.

This resulted in 3514 victims. That's going to take like 90 hours of run-time to scrape, if I do it the same way as I did with Seguin/Rigby... Idk about that.

Here's the build_chron_am_search function I used beforehand:

In [7]:
def build_chron_am_search(row):
    base_url = "https://chroniclingamerica.loc.gov/search/pages/results/list/"
    date1 = row['Year']
    date2 = row['Year'] + 1
    phrasetext = row['Name'].replace(' ', '+')

    search_url = (f'{base_url}?date1={date1}&date2={date2}&searchType=advanced&language='
                  f'&proxdistance=5&rows=1000&ortext=&proxtext=&phrasetext={phrasetext}'
                  f'&andtext=&dateFilterType=yearRange&page=1&sort=date')
    
    return search_url

black_victims['search_url'] = black_victims.apply(build_chron_am_search, axis=1)

black_victims

Status  Year  Month Day              Name        Alt Name #1  \
6              Lynching  1865      7  20   charles winters                NaN   
7              Lynching  1865      7  20      thos bradley                NaN   
8              Lynching  1865      7  21        john hirst                NaN   
9              Lynching  1865      7  21    john middleton                NaN   
22             Lynching  1866      2  16      joseph balls       Joseph Rails   
...                 ...   ...    ...  ..               ...                ...   
5445           Lynching  1921     12   4         west hale           Wes Hale   
5446           Lynching  1921     12  11        fred rouse           Ed Rouse   
5447  Probable lynching  1921     12  11    lonnie newsome     Alonzo Newsome   
5449           Lynching  1921     12  16  lee robinson jr   Joe Robertson Jr.   
5452           Lynching  1921     12  28   bill mcallister               Ed —   

           Alt Name #2 Alt Name #3   Order Victim's Race  ... Problem Flag  \
6                  NaN         NaN  2 of 2         Black  ...          NaN   
7                  NaN         NaN  1 of 2         Black  ...          NaN   
8                  NaN         NaN  1 of 2         Black  ...          NaN   
9                  NaN         NaN  2 of 2         Black  ...          NaN   
22        Joseph Ralls         NaN  1 of 2         Black  ...          NaN   
...                ...         ...     ...           ...  ...          ...   
5445               NaN         NaN  3 of 3         Black  ...          NaN   
5446               NaN         NaN  1 of 1         Black  ...          NaN   
5447               NaN         NaN  1 of 1         Black  ...          NaN   
5449  Leo Robinson Jr.         NaN  1 of 1         Black  ...          NaN   
5452               NaN         NaN  1 of 1         Black  ...          NaN   

                                        Resident Status  Time of Lynching  \
6     No indication of being a stranger or marginalized             Night   
7     No indication of being a stranger or marginalized             Night   
8     No indication of being a stranger or marginalized             Night   
9     No indication of being a stranger or marginalized             Night   
22    No indication of being a stranger or marginali...             Night   
...                                                 ...               ...   
5445  No indication of being a stranger or marginalized               NaN   
5446  No indication of being a stranger or marginali...          11:00 PM   
5447  No indication of being a stranger or marginalized             Night   
5449  No indication of being a stranger or marginali...               NaN   
5452  No indication of being a stranger or marginali...             Night   

                                             TolnayBeck        Torture  \
6     Original Tolnay-Beck inventory covers the year...  None reported   
7     Original Tolnay-Beck inventory covers the year...  None reported   
8     Original Tolnay-Beck inventory covers the year...  None reported   
9     Original Tolnay-Beck inventory covers the year...  None reported   
22    Original Tolnay-Beck inventory covers the year...  None reported   
...                                                 ...            ...   
5445  West Hale, 12/4/1921, Oconee Co., Black Male, ...  None reported   
5446  Original Tolnay-Beck inventory did not include...         Beaten   
5447  Original Tolnay-Beck inventory did not include...  None reported   
5449  Lee Robinson, Jr., 12/21/1921, Brooks Co., Bla...  None reported   
5452  Unnamed Negro, 1/4/1922, Florence Co., Black M...  None reported   

      Torture Codes                                       Verification  \
6                 0  Dayton (OH) Daily Empire 8/7/1865, p. 3 (Chron...   
7                 0  Dayton (OH) Daily Empire 8/7/1865, p. 3 (Chron...   
8                 0  Dayton (OH) Daily Empire 8/7/1865, p. 3 (Chron... 

In [8]:
black_victims.to_csv('black_subset_tolnay_beck.csv')

Okay, I'm all set to scrape the Tolnay-Beck results using my scrape_carefully() function and the other code in my build_refine_dataset.ipynb. But I'm gonna hold off because scraping this much from Chron Am is going to take some planning. I could break the dataset apart into more manageable chunks and scrape them over the course of a couple weeks?

Here's one way to do that, starting by segmenting the data into separate files:

In [11]:
victim_segments = np.array_split(black_victims, len(black_victims) // 300)

for i, segment in enumerate(victim_segments):
    segment_file = f'black_victims_segment_{i + 1}.csv'
    segment.to_csv(f'black_victim_data_segments/{segment_file}', index=False)

/opt/anaconda3/envs/nlp/lib/python3.11/site-packages/numpy/core/fromnumeric.py:59: FutureWarning: 'DataFrame.swapaxes' is deprecated and will be removed in a future version. Please use 'DataFrame.transpose' instead.
  return bound(*args, **kwds)


Then run scrape_carefully() and other code found in my build_refine_dataset.ipynb over each segment, probably one per night.